packages/cubejs-backend-shared/src/time.ts

```diff
@@ -26,10 +31,113 @@ export const TIME_SERIES: Record<string, (range: DateRange, timestampPrecision:
 };
 
-type TimeSeriesOptions = {
-  timestampPrecision: number
+/**
+ * Parse PostgreSQL-like interval string into object
+ * E.g. '2 years 15 months 100 weeks 99 hours 15 seconds'
+ * Negative units are also supported
+ * E.g. '-2 months 5 days -10 hours'
+ */
+export function parseSqlInterval(intervalStr: SqlInterval): ParsedInterval {
+  const interval: ParsedInterval = {};
+  const parts = intervalStr.split(/\s+/);
+
+  for (let i = 0; i < parts.length; i += 2) {
+    const value = parseInt(parts[i], 10);
+    const unit = parts[i + 1];
+
+    // Remove ending 's' (e.g., 'days' -> 'day')
+    const singularUnit = (unit.endsWith('s') ? unit.slice(0, -1) : unit) as unitOfTime.DurationConstructor;
+    interval[singularUnit] = value;
+  }
+
+  return interval;
+}
+
+export function addInterval(date: moment.Moment, interval: ParsedInterval): moment.Moment {
+  const res = date.clone();
+
+  Object.entries(interval).forEach(([key, value]) => {
+    res.add(value, key as unitOfTime.DurationConstructor);
+  });
+
+  return res;
+}
+
+export function subtractInterval(date: moment.Moment, interval: ParsedInterval): moment.Moment {
+  const res = date.clone();
+
+  Object.entries(interval).forEach(([key, value]) => {
+    res.subtract(value, key as unitOfTime.DurationConstructor);
+  });
+
+  return res;
+}
+
+/**
+ * Returns the closest date prior to date parameter aligned with the origin point
+ */
+function alignToOrigin(startDate: moment.Moment, interval: ParsedInterval, origin: moment.Moment): moment.Moment {
+  let alignedDate = startDate.clone();
+  let intervalOp;
+  let isIntervalNegative = false;
+
+  let offsetDate = addInterval(origin, interval);
+
+  // The easiest way to check the interval sign
+  if (offsetDate.isBefore(origin)) {
+    isIntervalNegative = true;
+  }
+
+  offsetDate = origin.clone();
+
+  if (startDate.isBefore(origin)) {
+    intervalOp = isIntervalNegative ? addInterval : subtractInterval;
+
+    while (offsetDate.isAfter(startDate)) {
+      offsetDate = intervalOp(offsetDate, interval);
+    }
+    alignedDate = offsetDate;
+  } else {
+    intervalOp = isIntervalNegative ? subtractInterval : addInterval;
+
+    while (offsetDate.isBefore(startDate)) {
+      alignedDate = offsetDate.clone();
+      offsetDate = intervalOp(offsetDate, interval);
+    }
+
+    if (offsetDate.isSame(startDate)) {
+      alignedDate = offsetDate;
+    }
+  }
+
+  return alignedDate;
+}
+
+export const timeSeriesFromCustomInterval = (intervalStr: string, [startStr, endStr]: QueryDateRange, origin: moment.Moment, options: TimeSeriesOptions = { timestampPrecision: 3 }): QueryDateRange[] => {
+  const intervalParsed = parseSqlInterval(intervalStr);
+  const start = moment(startStr);
+  const end = moment(endStr);
+  let alignedStart = alignToOrigin(start, intervalParsed, origin);
+
+  const dates: QueryDateRange[] = [];
+
+  while (alignedStart.isBefore(end)) {
```

Comment:
I am thinking about protection for super large time series. for example, user can specify 2 seconds interval from 1970 to 2050 and it will do a CPU spike (I don't think that manipulation with dates are fast + OOM?)

I belive, we should care about it. It's vector for DDoS

Reply:
How can you distinguish between malicious and legitimate requests? I think it's not possible.

Reply:
By limiting via configuration, we should throw an error when the size of the times series reaches N. сС @paveltiunov

Reply:
Users could use `second` or `minute` as a granularity for many years and I don't recall I single complaint about degraded performance due to this. So I'm not sure if we should have safeguards for such use cases.

I strongly feel we should not add an env var a a config option that we'll need to document, etc. If we really wanna do that, maybe it could be just a warning if the size of time series is more than N, and we just choose that N ourselves.

But I would prefer we just don't do anything here.

Reply:
Made a hard limit (50000) for now, and left a todo to make it configurable. This will be done in the one of upcoming PRs.